### PR TITLE
[Inductor] use two step algorithm to calculate variance

### PR DESCRIPTION
**Pull request by yuchengliu1**
Two step algorithm has a better performance when the reduction size is small compared to Welford algorithm. 
And when the size is small two step algorithm can mitigate "big - big = small" in reduction and focus on "big - small". 
Thus we tune a threshold, and the size below the threshold will gain the performance benefit and will not introduce precision problem.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -755,6 +755,9 @@ def use_autoheuristic(name: str) -> bool:
 # replace small reductions with pointwise, disable with `= 1`
 unroll_reductions_threshold = 8
 
+# use two step variance in small sizes, disable two step variance with `= 0`
+two_step_variance_threshold = 25600
```
Comment (CaoE) [inline torch/_inductor/config.py]: Unless absolutely necessary, it's better not to add more options here.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -6248,7 +6248,7 @@ def use_two_step_variance(x, axis, keepdim):
     reduction_numel = sympy_product(kwargs["reduction_ranges"])
     return (
         isinstance(reduction_numel, sympy.Integer)
-        and int(reduction_numel) < config.unroll_reductions_threshold
+        and (config.two_step_variance_threshold>0 and int(reduction_numel) <= config.two_step_variance_threshold)
```
Comment (CaoE) [inline torch/_inductor/lowering.py]: It is better to handle the dynamic shape case.

File: test/inductor/test_cpu_repro.py [hunk 3]
```diff
@@ -2759,6 +2759,33 @@ def fn(x):
...
+            def __init__(self, normalized_shape=N, eps=1e-5):
+                super(L, self).__init__()
+                self.layernorm = torch.nn.LayerNorm(
+                    normalized_shape, eps=eps)
+
+            def forward(self, x):
+                return self.layernorm(x)
+        mod = L().eval()
+        for mean in [0, 1e10]:
+            for std in [10, 1e10]:
```
Comment (CaoE) [inline test/inductor/test_cpu_repro.py]: This might make the input value too large, preventing the precision issues from being fully exposed. Could we add a more granular factor?

File: torch/_inductor/config.py (hunk 1 above)
Comment (yuchengliu1) [reply torch/_inductor/config.py]: removed

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (yuchengliu1) [reply torch/_inductor/lowering.py]: Dynamic shape should not use two step algorithm.  Current code can handle this.

File: test/inductor/test_cpu_repro.py (hunk 3 above)
Comment (yuchengliu1) [reply test/inductor/test_cpu_repro.py]: add randn(0,1)

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -6298,9 +6298,10 @@ def use_two_step_variance(x, axis, keepdim):
 
     ranges = kwargs["ranges"]
     reduction_numel = sympy_product(kwargs["reduction_ranges"])
+    two_step_variance_threshold = 1024
```
Comment (CaoE) [inline torch/_inductor/lowering.py]: Is it better to use a param in `use_two_step_variance` ?
```
def use_two_step_variance(x, axis, keepdim, threshold=1024):
...
    return (
        isinstance(reduction_numel, sympy.Integer)
        and int(reduction_numel) <= threshold
        and sympy_product(ranges) != 1
    )

```

File: test/inductor/test_cpu_repro.py [hunk 5]
```diff
@@ -2776,6 +2778,34 @@ def fn(x):
...
+            row_stds = x.std(dim=1, keepdim=True, unbiased=True)
+            x_norm = (x - row_means) / (row_stds + 1e-5)
+            x = x_norm * std + mean
+            input = (x,)
+            output_eager = mod(*input)
+            with torch.no_grad():
+                m = torch.compile(mod)
+                output_compiled = m(*input)
+            self.assertTrue(
+                torch.allclose(output_eager, output_compiled, atol=1, rtol=1e-4)
```
Comment (CaoE) [inline test/inductor/test_cpu_repro.py]: Set different `atol`/`rtol` for different mean/std. atol=1 is too large for small mean/std.

File: torch/_inductor/lowering.py [hunk 6]
```diff
@@ -6289,7 +6289,7 @@ def var_mean_sum_(x, axis, correction, keepdim, return_mean):
     return x_var, x_mean
 
 
-def use_two_step_variance(x, axis, keepdim):
+def use_two_step_variance(x, axis, keepdim, threshold=1024):
```
Comment (mingfeima) [inline torch/_inductor/lowering.py]: how do we get this magic number `1024` here? any benchmarks to demonstrate the the perf benefits and accuracy loss?

Comment (yuchengliu1) [reply torch/_inductor/lowering.py]: Two step algorithm has perf benefits in small amounts of data compared to Welford. However it has lower accuracy because of error accumulate. 
I benchmark it on CPU and found two step algorithm has a better perf when n<25600 on CPU. And 1024 can avoid the error accumulating to large,

Comment (yuchengliu1) [reply torch/_inductor/lowering.py]: update in comment

File: test/inductor/test_cpu_repro.py [hunk 7]
```diff
@@ -2776,6 +2778,39 @@ def fn(x):
...
+            row_means = x.mean(dim=1, keepdim=True)
+            row_stds = x.std(dim=1, keepdim=True, unbiased=True)
+            x_norm = (x - row_means) / (row_stds + 1e-5)
+            x = x_norm * std + mean
+            input = (x,)
+            output_eager = mod(*input)
+            with torch.no_grad():
+                m = torch.compile(mod)
+                output_compiled = m(*input)
+            if mean == 1e10 or std == 1e10:
```
Comment (CaoE) [inline test/inductor/test_cpu_repro.py:2850]: Nit:
```
            atol=1e-4
            rtol=1e-4
            if mean == 1e10 or std == 1e10:
                 atol=1

            self.assertTrue(torch.allclose(output_eager, output_compiled, atol=atol, rtol=rtol))
```

**Review by CaoE — DISMISSED** [review]

File: test/inductor/test_cpu_repro.py (hunk 7 above)
Comment (CaoE) [reply test/inductor/test_cpu_repro.py:2850]: We can add some more options:
`for mean, std in [(0, 1e10), (1e10, 10), (1e10, 1e10), (0, 1), (0, 0.5), (0.5, 1), (0, 2)]:`

Comment (yuchengliu1) [reply test/inductor/test_cpu_repro.py:2850]: done

**Review by mingfeima — DISMISSED** [review]

**Comment by yuchengliu1** [comment]
Hi @peterbell10 @jansel , could you kindly help to take a look of this PR?

**Comment by Valentine233** [comment]
@peterbell10 @jansel Could you please review this PR? Thanks!

File: torch/_inductor/lowering.py (hunk 6 above)
Comment (jansel) [reply torch/_inductor/lowering.py]: Let's move this 1024 magic number to cpu section of _inductor.config.

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — CHANGES_REQUESTED** [review]
Something is off with CI, please rebase and get tests green.

**Comment by yuchengliu1** [comment]
@jansel  I have moved `threshold` to config. The failures in CI  is unrelated to this PR. please review this PR again

File: torch/_inductor/config.py [hunk 8]
```diff
@@ -835,6 +835,10 @@ def use_autoheuristic(name: str) -> bool:
 # replace small reductions with pointwise, disable with `= 1`
 unroll_reductions_threshold = 8
 
+# threshold between two step reduction algorithm and welford reduction algorithm
+# only activate on CPU now
+two_step_var_threshold = 1024
```
Comment (jansel) [inline torch/_inductor/config.py]: Move the CPU-specific stuff to `class cpp:` full words for config names

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by jansel — CHANGES_REQUESTED** [review]
Fix lint error

**Comment by yuchengliu1** [comment]
The lint error is not because of my code but caused by the CI env. The error is `flake8.exceptions.FailedToLoadPlugin: Flake8 failed to load plugin "flake8-logging-format`.

**Review by jansel — DISMISSED** [review]

**Comment by yuchengliu1** [comment]
@jansel  I have fix the bug. Could you review this PR to reland.

**Review by mingfeima — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

**Comment by yuchengliu1** [comment]
@jansel  The New Failure is out of time limit. This failure has been occurring frequently lately.
